okio/jvm/src/test/java/okio/PipeKotlinTest.kt

```diff
@@ -32,4 +49,682 @@ class PipeKotlinTest {
+    pipeSink.emit()
+    assertEquals("world", actualSinkBuffer.readUtf8(5))
+
+    assertFailsWith<IllegalStateException> {
+      pipeSource.readUtf8()
+    }
+
+    pipeSink.close()
+    assertTrue(actualSinkClosed)
+  }
+
+  @Test
+  fun foldWritesPipeContentsToSink() {
+    val pipe = Pipe(128)
+
+    val pipeSink = pipe.sink.buffer()
+    pipeSink.writeUtf8("hello")
+    pipeSink.emit()
+
+    val foldSink = Buffer()
+    pipe.fold(foldSink)
+
+    assertEquals("hello", foldSink.readUtf8(5))
+  }
+
+  @Test
+  fun foldUnblocksBlockedWrite() {
+    val pipe = Pipe(4)
+    val foldSink = Buffer()
+
+    val latch = CountDownLatch(1)
+    executorService.schedule({
+      pipe.fold(foldSink)
+      latch.countDown()
+    }, 500, TimeUnit.MILLISECONDS)
+
+    val sink = pipe.sink.buffer()
+    sink.writeUtf8("abcdefgh") // Blocks writing 8 bytes to a 4 byte pipe.
+    sink.close()
+
+    latch.await()
+    assertEquals("abcdefgh", foldSink.readUtf8())
+  }
+
+  @Test
+  fun accessSinkAfterFold() {
+    val pipe = Pipe(100L)
+    pipe.fold(Buffer())
+    assertFailsWith<IllegalStateException> {
+      pipe.source.read(Buffer(), 1L)
+    }
+  }
+
+  @Test
+  fun honorsPipeSinkTimeoutOnWritingWhenItIsSmaller() {
+    val smallerTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(500L)
+    val biggerTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(1500L)
+
+    val pipe = Pipe(4)
+    val underlying = TimeoutWritingSink()
+
+    underlying.timeout.timeout(biggerTimeoutNanos, TimeUnit.NANOSECONDS)
+    pipe.sink.timeout().timeout(smallerTimeoutNanos, TimeUnit.NANOSECONDS)
+
+    pipe.fold(underlying)
+
+    val start = System.currentTimeMillis()
+    pipe.sink.write(Buffer().writeUtf8("abc"), 3)
+    val elapsed = TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis() - start)
+
+    assertEquals(smallerTimeoutNanos.toDouble(), elapsed.toDouble(),
+      TimeUnit.MILLISECONDS.toNanos(100).toDouble())
+    assertEquals(biggerTimeoutNanos, underlying.timeout().timeoutNanos())
+  }
+
+  @Test
+  fun honorsUnderlyingTimeoutOnWritingWhenItIsSmaller() {
+    val smallerTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(500L)
+    val biggerTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(1500L)
+
+    val pipe = Pipe(4)
+    val underlying = TimeoutWritingSink()
+
+    underlying.timeout.timeout(smallerTimeoutNanos, TimeUnit.NANOSECONDS)
+    pipe.sink.timeout().timeout(biggerTimeoutNanos, TimeUnit.NANOSECONDS)
+
+    pipe.fold(underlying)
+
+    val start = System.currentTimeMillis()
+    pipe.sink.write(Buffer().writeUtf8("abc"), 3)
+    val elapsed = TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis() - start)
+
+    assertEquals(smallerTimeoutNanos.toDouble(), elapsed.toDouble(),
+      TimeUnit.MILLISECONDS.toNanos(100).toDouble())
+    assertEquals(smallerTimeoutNanos, underlying.timeout().timeoutNanos())
+  }
+
+  @Test
+  fun honorsPipeSinkTimeoutOnFlushingWhenItIsSmaller() {
+    val smallerTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(500L)
```

Comment:
I wonder if it’s worth having class-level fields so we don’t need to keep saying these two